sourceinfo/source_code_info.go

```diff
@@ -580,43 +613,95 @@ func (sci *sourceCodeInfo) getTrailingComments(n ast.Node) comments {
 	}
 	info := sci.file.TokenInfo(e)
 	nextInfo := sci.file.TokenInfo(next)
-	c, _ := attributeComments(info, nextInfo)
-	return c
+	t, _, _ := sci.attributeComments(info, nextInfo)
+	return t
 }
 
-func attributeComments(prevInfo, info ast.NodeInfo) (t comments, l []comments) {
-	trail := prevInfo.TrailingComments()
-	lead := groupComments(info.LeadingComments())
-
-	if trail.Len() > 0 || len(lead) == 0 {
-		// previous token already has trailing comments or current token
-		// has no comments from which it may borrow
-		return trail, lead
+func (sci *sourceCodeInfo) attributeComments(prevInfo, info ast.NodeInfo) (t comments, d []comments, l comments) {
+	detached := groupComments(info.LeadingComments())
+	var trail comments
+	if prevInfo.IsValid() {
+		trail = comments(prevInfo.TrailingComments())
+		if trail.Len() == 0 {
+			trail, detached = sci.maybeDonate(prevInfo, info, detached)
+		}
+	} else {
+		trail = emptyComments
 	}
-	if prevInfo.End().Line == info.Start().Line {
-		// if the tokens are on the same line, no way to attribute
-		return trail, nil
+	detached, lead := sci.maybeAttach(prevInfo, info, trail.Len() > 0, detached)
+	return trail, detached, lead
+}
+
+func (sci *sourceCodeInfo) maybeDonate(prevInfo ast.NodeInfo, info ast.NodeInfo, lead []comments) (t comments, l []comments) {
+	if len(lead) == 0 {
+		// nothing to donate
+		return emptyComments, nil
+	}
+	if !sci.extraComments {
+		// Mirroring protoc for now: if tokens on the same line, attribution
+		// is ambiguous so drop everything.
+		// This condition can be dropped (and we can stop dropping comments)
+		// once this PR is released:
+		//   https://github.com/protocolbuffers/protobuf/pull/10660
+		if prevInfo.End().Line == info.Start().Line {
+			return emptyComments, nil
+		}
 	}
 	if lead[0].Index(0).Start().Line > prevInfo.End().Line+1 {
 		// first comment is detached from previous token, so can't be a trailing comment
-		return trail, lead
+		return emptyComments, lead
 	}
 	if len(lead) > 1 {
 		// multiple groups? then donate first comment to previous token
 		return lead[0], lead[1:]
 	}
-
-	if lead[0].Index(lead[0].Len()-1).End().Line < info.Start().Line-1 {
+	// there is only one element in lead
+	comment := lead[0]
+	if comment.Index(comment.Len()-1).End().Line < info.Start().Line-1 {
 		// there is a blank line between the comments and subsequent token, so
 		// we can donate the comment to previous token
-		return lead[0], lead[1:]
+		return comment, nil
 	}
-	if txt := info.RawText(); len(txt) == 1 && strings.Contains("}])", txt) {
+	if txt := info.RawText(); len(txt) == 1 && strings.Contains("}]),;", txt) {
 		// token is a symbol for the end of a scope, which doesn't need a leading comment
-		return lead[0], lead[1:]
+		return comment, nil
+	}
+
+	// cannot donate
+	return emptyComments, lead
+}
+
+func (sci *sourceCodeInfo) maybeAttach(prevInfo ast.NodeInfo, info ast.NodeInfo, hasTrail bool, lead []comments) (d []comments, l comments) {
```

Comment:
I moved this step out of `newLocWithGivenComments` because it now needs to know location info for the token to decide how if a comment's attachment is ambiguous. (This wasn't previously an issue because, when not trying to collect _all_ comments, we intentionally _drop_ such ambiguous comments in order to mirror the behavior of protoc.)